### PR TITLE
[WIP] Refactoring Envoy process management

### PR DESCRIPTION
**Pull request by nmittler**
The existing envoy agent/proxy code is monolithic, making it difficult to extend, reuse parts, as well as to generally understand.

This PR attempts to simplify the logic for managing the lifecycle of Envoy processes. The new code is protected by an environment variable, which is disabled by default.

A few key changes/additions:

- The restart.Manager API now supports providing handlers for lifecycle events for Envoy and the manager, itself.

- When the Manager is created, it automatically attempts to start Envoy for the first time. The old code was waiting for the first cert file update before starting Envoy. We address this in the pilot_agent code, by waiting for the first cert file update before creating the manager.

- Explicitly document that when the run context is cancelled, that a graceful shutdown (via draining) is initiated.

- If an Envoy process terminates with no error, attempts to restart Envoy. It was unclear (and possibly a bug) that we weren't doing this before.

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/cmd/pilot-agent/main.go [hunk 1]
```diff
@@ -413,12 +422,72 @@ var (
...
-			envoyProxy := envoy.NewProxy(proxyConfig, role.ServiceNode(), proxyLogLevel, proxyComponentLogLevel, pilotSAN, role.IPAddresses, dnsRefreshRate, opts)
-			agent := envoy.NewAgent(envoyProxy, envoy.DefaultRetry, features.TerminationDrainDuration())
-			watcher := envoy.NewWatcher(tlsCertsToWatch, agent.ConfigCh())
+			// TODO(nmittler): Once the new Envoy manager has baked, get rid of the variable and the old code.
+			if useNewEnvoyManager.Get() {
+				// Create a function for running the Envoy restart manager.
+				runEnvoyManager := func() restart.Manager {
+					mgr := restart.New(restart.Config{
+						MaxRestartAttempts:       uint(envoy.DefaultRetry.MaxRetries),
+						InitialRestartInterval:   envoy.DefaultRetry.InitialInterval,
```
Comment (howardjohn) [inline pilot/cmd/pilot-agent/main.go]: this is maybe outside the scope of this PR but related -- I am starting to think this whole restarting on envoy crash with backoff etc is just shooting ourselves in the foot for no reason. Consider a case where Envoy is going OOM.
* Envoy dies, but kubernetes still reports things are good and 2/2 ready. Repeat a few times.
* At somepoint due to backoff, envoy is dead for 2 minutes. The readiness probe won't fail for at least a minute, so there is 1min where traffic is coming and we are not even attempting to handle it.
* Meanwhile, the operator has no clue what is going on, because Kubernetes doesn't know envoy is crashing

Alternatively, if we exit when envoy crashes:
* The readiness probe fails immediately
* The user can clearly see envoy is crashing
* Kubernetes handles backoff for us
* We don't have to handle the complexity.

We do still need hot restarts, but I think that is orthogonal to restarting failed Envoy's

File: pkg/envoy/restart/handler.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+func (h *handlerImpl) OnStarting() {
+	for _, f := range h.onStarting {
+		f()
+	}
+}
+
+func (h *handlerImpl) OnShuttingDown() {
+	for _, f := range h.onShuttingDown {
+		f()
```
Comment (howardjohn) [inline pkg/envoy/restart/handler.go]: do we need this abstraction here? I tend to find this impossible to figure out what is actually being called in set ups like this. Can we just do 
```
func OnStarting() {
   startfn1()
   startfn2()
}
```
and keep it explicit? 

I don't see where this is actually being called either (probably just github being annoying)

File: pilot/cmd/pilot-agent/main.go (hunk 1 above)
Comment (nmittler) [reply pilot/cmd/pilot-agent/main.go]: That's a good point.  Seems reasonable to me.

@kyessenov @costinm @rshriram  WDYT?

Comment (kyessenov) [reply pilot/cmd/pilot-agent/main.go]: The restart was addressing races in hot-restart protocol and not permanent failures (like bad bootstrap). We can't really distinguish the two, and we know that hot restart has race conditions.

File: pkg/envoy/restart/handler.go (hunk 2 above)
Comment (kyessenov) [reply pkg/envoy/restart/handler.go]: +1. Please don't invert the control flow, it makes very much unreadable.

File: pilot/cmd/pilot-agent/main.go (hunk 1 above)
Comment (nmittler) [reply pilot/cmd/pilot-agent/main.go]: If we're using SDS can we simply turn off restarts altogether? In the termination logic, we abort all processes created by the manager, which *should* resolve hot-restart races IIUC.

File: pkg/envoy/restart/handler.go (hunk 2 above)
Comment (nmittler) [reply pkg/envoy/restart/handler.go]: This is a composite handler, so each event has `n` handler functions that need be notified. All we're doing here is iterating through the appropriate event handlers.

If the concern is simply the idea of a composite handler, I'm fine removing it for now.

sg?

Comment (kyessenov) [reply pkg/envoy/restart/handler.go]: I just want to be able to read the follow the code by following the function calls. Every time there's some function handler that is assigned separately, it makes it harder to follow.

File: pilot/cmd/pilot-agent/main.go (hunk 1 above)
Comment (kyessenov) [reply pilot/cmd/pilot-agent/main.go]: Hot restart race happens when a starting envoy talks to a running envoy, and there's some mis-understanding over the protocol, which usually causes the new envoy to crash. If I remember, we found at least 3 segfaults falling in this category, and still not clear if any are left. Usually, another attempt helps to resolve since the running envoy can re-negotiate handling over the sockets.

Comment (howardjohn) [reply pilot/cmd/pilot-agent/main.go]: @kyessenov my understanding is if we intentionally trigger a hot restart (certs), there is no traffic loss. But if Envoy crashes, its no longer a hot restart -- its just a normal restart, with normal traffic lost. In this case why do we need pilot-agent to restart envoy, rather than just having kubernetes do it? I think if kubernetes does it it will immediately get marked as unready as well which has added benefit.

File: pkg/envoy/restart/handler.go (hunk 2 above)
Comment (nmittler) [reply pkg/envoy/restart/handler.go]: So the requirement here is to allow the caller to provide independent handler for lifecycle events, so I'm not that we can eliminate some abstraction here.

Comment (howardjohn) [reply pkg/envoy/restart/handler.go]: I am not sure I understand. We aren't building a generic library here that will be used by many users with arbitrary call goals -- aren't we are the only users of this? Can we just replace this with 

```go
func OnStarting() {
			m.eventCh <- Event{
				Type: Starting,
			}
}
```

(and if thats not actually the code that is called from OnStarting then that is another sign that its a bit hard to follow)

File: pkg/envoy/proxy.go [hunk 3]
```diff
@@ -41,17 +40,18 @@ const (
...
+	extraArgs             []string
+	pilotSAN              []string
+	opts                  map[string]interface{}
+	nodeIPs               []string
+	dnsRefreshRate        string
 }
 
 // NewProxy creates an instance of the proxy control commands
-func NewProxy(config meshconfig.ProxyConfig, node string, logLevel string,
+func NewProxy(config meshconfig.ProxyConfig, bootstrapOverrideFile string, node string, logLevel string,
```
Comment (liamawhite) [inline pkg/envoy/proxy.go]: Should we use functional options here? There are a lot of parameters!

File: pkg/envoy/restart/handler.go (hunk 2 above)
Comment (liamawhite) [reply pkg/envoy/restart/handler.go]: IMO, the lack of extensibility here is how we ended up with the agent kludge we have today. There's no clear separation of what's core to running Envoy and what are extensions. As a result, pilot-agent expands to be the tightly coupled monolith it is today. This PR moves towards a more composable and modular pilot-agent.

Comment (liamawhite) [reply pkg/envoy/restart/handler.go]: The implementation details of how we do this should definitely be discussed though!

Comment (howardjohn) [reply pkg/envoy/restart/handler.go]: If we want to make it extensible then it seems we could do stuff like

```go
func OnStarting() {
   startEnvoy()
   startMonitoring()
   startMaintenance()
}
```

if I am understanding the goals here correctly?

I think we can make it clean without being this abstract. Its really hard to tell what is going on here. I am still not 100% sure what exactly is called from `OnStarting` even after pulling it into my IDE, and similar patterns across the codebase have been a challenge to debug in the past.

Comment (liamawhite) [reply pkg/envoy/restart/handler.go]: @kyessenov What were the upstreaming requirements you had in the preceding PR?

File: pilot/cmd/pilot-agent/main.go (hunk 1 above)
Comment (nmittler) [reply pilot/cmd/pilot-agent/main.go]: I think leaving restarts to k8s makes sense.  If we built hot restart into envoy.Instance, we could effectively get rid of this manager altogether. 

@kyessenov are we missing other cases for why pilot-agent is managing the restarts? VMs perhaps?

Comment (liamawhite) [reply pilot/cmd/pilot-agent/main.go]: `systemd` manages restarts on VMs I think?

**Comment by kyessenov** [comment]
The problem with applying static type here is that we don't control the declarations. These are a part of xDS APIs and change frequently. Putting static types in place makes adapting to changes in xDS  harder, especially since we don't follow cross-versioning systematically (e.g. which Envoys would istio 1.3.1 work with?)
